### ISSUE TITLE
Cognitive Memory System: Procedural Memory and Cross-Project Learning Are Unimplemented

### ISSUE THREAD

**Issue by lwgray** [lwgray/marcus#195]
## Summary

The Memory system (`src/core/memory.py`) is built on a four-tier cognitive model — Working, Episodic, Semantic, and Procedural — but only ~60% of it is functional. Two of the four tiers work well, one is partially implemented, and one is entirely scaffolding. Cross-project learning doesn't exist. This issue documents the gaps and proposes a path forward.

Related: #176 (Close the Learning Loop) covers the broader problem of memory insights not flowing back into decisions. This issue focuses specifically on the **incomplete memory infrastructure** that #176 would build on.

---

### ✅ Working Memory — Functional
- Tracks active tasks per agent, recent events, system state
- Cleared on task completion
- Used throughout the system

### ✅ Episodic Memory — Functional
- `record_task_completion()` stores `TaskOutcome` objects with actual hours, blockers, success/failure
- Timeline populated by date
- Feeds prediction methods (`predict_task_outcome`, `predict_completion_time`, etc.)

### ⚠️ Semantic Memory — Partially Functional

**Working:**
- **Agent profiles** — `_update_agent_profile()` maintains skill success rates via exponential moving average, estimation accuracy, blocker history. Persisted. Actively used by `predict_task_outcome()`.
- **Task patterns** — `_learn_task_patterns()` builds patterns keyed by sorted labels, tracking durations, success rates, and best agents. Used by predictions.

**Not working:**
- **`success_factors`** — initialized as `{}`, never written to by any method. Dead field.

### ❌ Procedural Memory — Not Implemented

Initialized with three empty dicts in `__init__`:

```python
self.procedural: Dict[str, Any] = {
    "workflows": {},      # workflow_id -> steps
    "strategies": {},     # situation -> strategy
    "optimizations": {},  # pattern -> optimization
}
```

**No method in `Memory` or `MemoryAdvanced` ever reads from or writes to `self.procedural`.** It appears in `get_memory_stats()` (reporting length of empty dicts) and nowhere else. The system currently learns *facts about what happened* but not *how to do things better*.

### ❌ Cross-Project Learning — Missing

Memory is scoped to a single Marcus server instance. Each new project starts with a blank memory. There's no mechanism to carry learned agent profiles, task patterns, or (eventually) procedural knowledge across experiments or projects.

### ❌ ML Predictions — Placeholder

```python
async def predict_with_ml(self, agent_id: str, task: Task) -> Dict[str, Any]:
    """Future: Use ML model for predictions (placeholder for now)."""
    return await self.predict_task_outcome_v2(agent_id, task)
```

---

## What Procedural Memory Should Do

Procedural memory is the most interesting unimplemented tier because it represents the shift from **passive observation** to **active strategy**. Where episodic memory says "this happened" and semantic memory says "this is generally true," procedural memory should say "in this situation, do this."

Concrete examples of what procedural memory could learn and apply:

### Workflow Learning
- After N projects, Marcus observes that tasks labeled `authentication` + `third-party` succeed 40% more often when decomposed into 3 subtasks (research, implement, test) rather than 5.
- Observation gets encoded as a workflow: "For auth integration tasks, prefer 3-subtask decomposition with explicit research phase."
- The decomposer (`coordinator/decomposer.py`) consults procedural memory before calling the LLM for decomposition, injecting learned preferences into the prompt.

### Strategy Learning
- Marcus observes that when Agent X reports a blocker on a dependency, reassigning the blocking task to a higher-priority agent resolves it 3x faster than waiting.
- Encoded as a strategy: "When a blocker is reported on a dependency, escalate the blocking task's priority and consider reassignment."
- The blocker handler (`report_blocker` in `tools/task.py`) consults procedural memory for resolution strategies.

### Optimization Learning
- Marcus observes that cross-parent dependency wiring produces better parallelism when embedding similarity threshold is 0.5 instead of 0.6 for projects with fewer than 20 subtasks.
- Encoded as an optimization: "For small projects (<20 subtasks), lower embedding threshold to 0.5."
- The dependency wiring system reads this before running.

---

### Phase 1: Complete Semantic Memory
- Implement `success_factors` tracking — when a task succeeds, extract and weight the contributing factors (instruction style, decomposition depth, agent-task match score, context richness)
- This gives procedural memory something to learn *from*

### Phase 2: Procedural Memory — Recording
- Add `record_workflow()` — called after task/project completion to extract workflow patterns from episodic data
- Add `record_strategy()` — called when a decision (e.g., reassignment, priority change) leads to a measurably better outcome
- Add `record_optimization()` — called when a system parameter (threshold, timeout, decomposition depth) correlates with better results
- Persist to the same `Persistence` backend used by other tiers

### Phase 3: Procedural Memory — Retrieval
- Add `get_relevant_workflows(task_context)` — retrieves workflows that match the current situation
- Add `get_strategy(situation)` — retrieves learned strategies for a given situation type
- Add `get_optimizations(component)` — retrieves parameter optimizations for a system component
- Integrate into existing decision points: decomposer, task assignment, blocker resolution, dependency wiring

### Phase 4: Cross-Project Persistence
- Extract a `MemoryExporter` that serializes learned knowledge (agent profiles, task patterns, procedural entries) to a portable format
- Add a `MemoryImporter` that loads prior knowledge into a new project's memory, with decay factors for staleness
- This is what makes Marcus genuinely learn over time — each project benefits from all prior projects

### Phase 5: ML Integration (Optional)
- Replace the `predict_with_ml()` placeholder with an actual model trained on accumulated episodic data
- Could be as simple as a scikit-learn model trained on (task features, agent features) → (success probability, estimated duration)
- Or as sophisticated as a fine-tuned embedding model for better dependency wiring candidates

---

## Why This Matters for Marcus's Future

Marcus currently operates as a **stateless coordinator** — each project is a fresh start, each decision is made from scratch (or from within-session episodic data). The cognitive memory model was designed to change this, but only the observation layers are built.

The progression from the current state to a complete memory system represents Marcus's evolution from:

1. **Reactive coordinator** (current) — makes decisions based on current state + within-session history
2. **Learning coordinator** (Phase 1-3) — makes decisions informed by accumulated patterns and strategies
3. **Evolving coordinator** (Phase 4-5) — improves across projects, adapting its coordination strategies based on empirical outcomes

This is the same trajectory that separates a junior project manager (follows a process) from a senior one (has internalized patterns from hundreds of projects and adapts their approach to each new situation).

The dependency wiring system, the tiered instruction generator, the task decomposer — all of these already have injection points where procedural memory could inform decisions. The infrastructure is ready. The learning layer just needs to be built.
